File: Tests/cube.c
#include <stdio.h>
#include <windows.h>
#include "projectingAndRotating.h"

// Define the 8 vertices of a cube centered at (0,0,0)
// The cube has a side length of 20 units (-10 to 10)
Coords3d cubeVertices[8] = {
    {-10, -10, -10}, {10, -10, -10}, {10, 10, -10}, {-10, 10, -10},
    {-10, -10,  10}, {10, -10,  10}, {10, 10,  10}, {-10, 10,  10}
};

// Index pairs that define the 12 edges of a cube
int cubeEdges[12][2] = {
    {0, 1}, {1, 2}, {2, 3}, {3, 0}, // Back face edges
    {4, 5}, {5, 6}, {6, 7}, {7, 4}, // Front face edges
    {0, 4}, {1, 5}, {2, 6}, {3, 7}  // Connecting edges
};

int main() {
    // Canvas dimensions (your custom resolution)
    double centerX = 64.0;
    double centerY = 18.0;

    // Focal length / Camera distance
    // F is the viewing point. Z must be smaller than F.
    // If the cube is at Z=0, keeping F=60 means it's 60 units away from the camera.
    double F = 60.0; 

    double angleX = 0.0;
    double angleY = 0.0;

    // Array to store projected 2D points for rendering each frame
    // We will draw the 8 vertices + points along the edges
    int totalPointsToDraw = 8 + (12 * 5); // 8 vertices + 5 points per edge
    Coords2d pointsToDraw[100]; 

    // Hide cursor for a cleaner look
    printf("\033[?25l");

    while (1) {
        // 1. Clear the screen and buffer
        printf("\033[2J\033[3J\033[H");

        int pointIndex = 0;

        // 2. Rotate and project the 8 main vertices
        Coords3d rotatedVertices[8];
        for (int i = 0; i < 8; i++) {
            // Apply 3D rotation on multiple axes to make it spin dynamically
            Coords3d r = Rotate3DInX(cubeVertices[i].x, cubeVertices[i].y, cubeVertices[i].z, angleX);
            r = Rotate3DInY(r.x, r.y, r.z, angleY);
            rotatedVertices[i] = r;

            // Project 3D to 2D
            pointsToDraw[pointIndex++] = project(r.x, r.y, r.z, F);
        }

        // 3. Generate intermediate points along the edges to "draw" the lines
        for (int i = 0; i < 12; i++) {
            Coords3d p1 = rotatedVertices[cubeEdges[i][0]];
            Coords3d p2 = rotatedVertices[cubeEdges[i][1]];

            // Put 5 points between vertex 1 and vertex 2
            for (int j = 1; j <= 5; j++) {
                double t = j / 6.0; // Interpolation factor
                double ix = p1.x + (p2.x - p1.x) * t;
                double iy = p1.y + (p2.y - p1.y) * t;
                double iz = p1.z + (p2.z - p1.z) * t;

                pointsToDraw[pointIndex++] = project(ix, iy, iz, F);
            }
        }

        // 4. Render all points to your 128 x 36 terminal screen
    for (int i = 0; i < pointIndex; i++) {
        // Scaled down from 2.5 to 1.3 to make the cube smaller on screen
        double scale = 1.0; 

        int screenX = (int)(centerX + (pointsToDraw[i].x * 2.0 * scale));
        int screenY = (int)(centerY + (pointsToDraw[i].y * scale));

        // Boundary check to ensure we don't print off the screen
        if (screenX > 0 && screenX < 128 && screenY > 0 && screenY < 36) {
            printf("\033[%d;%dHo", screenY, screenX);
        }
    }

        fflush(stdout);

        // 5. Adjust rotation speeds for a mesmerizing tumbling effect
        angleX += 0.03;
        angleY += 0.04;

        Sleep(10); // Control frame rate (~25 FPS)
    }

    return 0;
}

File: rust/examples/cube.rs
//! Terminal spinning cube.
//!
//! Cross-platform Rust version of `Tests/cube.c` / `Tests/cubeLinux.c`.
//! Renders a wireframe cube using ANSI escape codes.

use std::io::{self, Write};
use std::thread;
use std::time::Duration;

use par::{Coords2d, Coords3d, project, rotate_3d_in_x, rotate_3d_in_y};

// ---- geometry -----------------------------------------------------------

/// 8 vertices of a cube centred at (0,0,0) with side length 20.
const CUBE_VERTICES: [Coords3d; 8] = [
    Coords3d {
        x: -10.0,
        y: -10.0,
        z: -10.0,
    },
    Coords3d {
        x: 10.0,
        y: -10.0,
        z: -10.0,
    },
    Coords3d {
        x: 10.0,
        y: 10.0,
        z: -10.0,
    },
    Coords3d {
        x: -10.0,
        y: 10.0,
        z: -10.0,
    },
    Coords3d {
        x: -10.0,
        y: -10.0,
        z: 10.0,
    },
    Coords3d {
        x: 10.0,
        y: -10.0,
        z: 10.0,
    },
    Coords3d {
        x: 10.0,
        y: 10.0,
        z: 10.0,
    },
    Coords3d {
        x: -10.0,
        y: 10.0,
        z: 10.0,
    },
];

/// 12 edges as index pairs into [`CUBE_VERTICES`].
const CUBE_EDGES: [(usize, usize); 12] = [
    (0, 1),
    (1, 2),
    (2, 3),
    (3, 0), // back face
    (4, 5),
    (5, 6),
    (6, 7),
    (7, 4), // front face
    (0, 4),
    (1, 5),
    (2, 6),
    (3, 7), // connecting edges
];

/// Interpolation points per edge (5 intermediate dots).
const POINTS_PER_EDGE: usize = 5;

// ---- render -------------------------------------------------------------

fn main() {
    let center_x = 64.0;
    let center_y = 18.0;
    let f = 60.0; // focal length / camera distance

    let mut angle_x = 0.0;
    let mut angle_y = 0.0;

    // Storage: 8 vertices + 12 edges × 5 interpolated points = 68 points.
    let mut pts = [Coords2d::default(); 128];
    let mut rotated = [Coords3d::default(); 8];

    // Hide cursor.
    print!("\x1b[?25l");
    let _ = io::stdout().flush();

    loop {
        // Clear screen + scrollback, return cursor to (1,1).
        print!("\x1b[2J\x1b[3J\x1b[H");
        let _ = io::stdout().flush();

        let mut n = 0;

        // ---- rotate & project the 8 vertices ----
        for i in 0..8 {
            let v = CUBE_VERTICES[i];
            let r = rotate_3d_in_x(v.x, v.y, v.z, angle_x);
            let r = rotate_3d_in_y(r.x, r.y, r.z, angle_y);
            rotated[i] = r;
            pts[n] = project(r.x, r.y, r.z, f);
            n += 1;
        }

        // ---- interpolate along edges ----
        for &(i1, i2) in &CUBE_EDGES {
            let p1 = rotated[i1];
            let p2 = rotated[i2];
            // The C original puts 5 evenly-spaced dots between the vertices.
            for j in 1..=POINTS_PER_EDGE {
                let t = j as f64 / (POINTS_PER_EDGE as f64 + 1.0);
                let ix = p1.x + (p2.x - p1.x) * t;
                let iy = p1.y + (p2.y - p1.y) * t;
                let iz = p1.z + (p2.z - p1.z) * t;
                pts[n] = project(ix, iy, iz, f);
                n += 1;
            }
        }

        // ---- blit ----
        for p in pts[..n].iter() {
            let sx = (center_x + p.x * 2.0) as i32;
            let sy = (center_y + p.y) as i32;
            if sx > 0 && sx < 128 && sy > 0 && sy < 36 {
                print!("\x1b[{};{}Ho", sy, sx);
            }
        }
        let _ = io::stdout().flush();

        // Advance rotation.
        angle_x += 0.03;
        angle_y += 0.04;

        thread::sleep(Duration::from_millis(40)); // ~25 FPS
    }
}
